### ISSUE TITLE
Problems restoring bower packages

### ISSUE THREAD

**Issue by tjrobinson** [Microsoft/vsts-tasks#5665]
## Environment
- Server: VSTS 
- Account name: empactis
- Team project name: Organisation Manager
- Build definition name/build number: https://empactis.visualstudio.com/Organisation%20Manager/Organisation%20Manager%20Team/_build/index?buildId=1869&_a=summary
- Agent: Hosted VS2017

## Issue Description
The restore of bower packages from our private GitHub repository is failing when it has been working previously. Examples below. I've run this build numerous times and keep getting this error or similar. Perhaps a networking issue?

### Error logs
```
2017-10-18T14:50:30.3210732Z ##[section]Starting: dotnet publish
2017-10-18T14:50:30.3219610Z ==============================================================================
2017-10-18T14:50:30.3219610Z Task         : .NET Core
2017-10-18T14:50:30.3219610Z Description  : Build, test and publish using dotnet core command-line.
2017-10-18T14:50:30.3219610Z Version      : 1.0.2
2017-10-18T14:50:30.3219610Z Author       : Microsoft Corporation
2017-10-18T14:50:30.3219610Z Help         : [More Information](https://go.microsoft.com/fwlink/?linkid=832194)
2017-10-18T14:50:30.3219610Z ==============================================================================
2017-10-18T14:50:49.1834772Z [command]"C:\Program Files\dotnet\dotnet.exe" publish d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj --configuration $(BuildConfiguration) --output d:\a\1\a\Empactis.OrganisationManager
2017-10-18T14:50:49.4934812Z Microsoft (R) Build Engine version 15.1.1012.6693
2017-10-18T14:50:49.5804814Z Copyright (C) Microsoft Corporation. All rights reserved.
2017-10-18T14:50:49.5804814Z 
2017-10-18T14:50:52.9604859Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Shared\Empactis.OrganisationManager.Shared.csproj]
2017-10-18T14:50:55.7584889Z   Empactis.OrganisationManager.Shared -> d:\a\1\s\src\Empactis.OrganisationManager.Shared\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Shared.dll
2017-10-18T14:50:56.2384740Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Data\Empactis.OrganisationManager.Data.csproj]
2017-10-18T14:50:59.9584575Z   Empactis.OrganisationManager.Data -> d:\a\1\s\src\Empactis.OrganisationManager.Data\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Data.dll
2017-10-18T14:51:00.6474315Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Commands\Empactis.OrganisationManager.Commands.csproj]
2017-10-18T14:51:04.2133806Z   Empactis.OrganisationManager.Commands -> d:\a\1\s\src\Empactis.OrganisationManager.Commands\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Commands.dll
2017-10-18T14:51:05.8493492Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj]
2017-10-18T14:51:09.5042721Z   Empactis.OrganisationManager -> d:\a\1\s\src\Empactis.OrganisationManager\bin\$(BuildConfiguration)\net462\win7-x86\Empactis.OrganisationManager.exe
2017-10-18T14:51:20.0985360Z   bower empactis-angular#1.14.0       not-cached https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:51:20.0985360Z   bower empactis-angular#1.14.0          resolve https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:51:20.9198380Z   bower empactis-angular#1.14.0         download https://github.com/empactis/empactis.angular/archive/1.14.0.tar.gz
2017-10-18T14:51:21.3168256Z   bower empactis-angular#1.14.0            retry Download of https://github.com/empactis/empactis.angular/archive/1.14.0.tar.gz failed with EHTTP, trying with git..
2017-10-18T14:51:21.3408244Z   bower empactis-angular#1.14.0         checkout 1.14.0
2017-10-18T14:51:23.4952332Z   bower empactis-angular#1.14.0         resolved https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:51:23.5332305Z   bower d3#~3.5.6                        EINVRES Request to https://bower.herokuapp.com/packages/d3 failed with 410
2017-10-18T14:51:23.5492311Z d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj(91,5): error MSB3073: The command "bower install" exited with code 1.
2017-10-18T14:51:23.6912251Z ##[error]Error: C:\Program Files\dotnet\dotnet.exe failed with return code: 1
2017-10-18T14:51:23.6912251Z ##[error]Dotnet command failed with non-zero exit code on the following projects : d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj
2017-10-18T14:51:23.7152237Z ##[section]Finishing: dotnet publish
```

```
2017-10-18T14:22:39.0095537Z ##[section]Starting: dotnet publish
2017-10-18T14:22:39.0095537Z ==============================================================================
2017-10-18T14:22:39.0095537Z Task         : .NET Core
2017-10-18T14:22:39.0095537Z Description  : Build, test and publish using dotnet core command-line.
2017-10-18T14:22:39.0095537Z Version      : 1.0.2
2017-10-18T14:22:39.0095537Z Author       : Microsoft Corporation
2017-10-18T14:22:39.0095537Z Help         : [More Information](https://go.microsoft.com/fwlink/?linkid=832194)
2017-10-18T14:22:39.0095537Z ==============================================================================
2017-10-18T14:22:54.1833295Z [command]"C:\Program Files\dotnet\dotnet.exe" publish d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj --configuration $(BuildConfiguration) --output d:\a\1\a\Empactis.OrganisationManager
2017-10-18T14:22:54.5993156Z Microsoft (R) Build Engine version 15.1.1012.6693
2017-10-18T14:22:54.6063202Z Copyright (C) Microsoft Corporation. All rights reserved.
2017-10-18T14:22:54.6063202Z 
2017-10-18T14:22:57.3592279Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Shared\Empactis.OrganisationManager.Shared.csproj]
2017-10-18T14:22:59.2561606Z   Empactis.OrganisationManager.Shared -> d:\a\1\s\src\Empactis.OrganisationManager.Shared\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Shared.dll
2017-10-18T14:22:59.7191428Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Data\Empactis.OrganisationManager.Data.csproj]
2017-10-18T14:23:02.1050518Z   Empactis.OrganisationManager.Data -> d:\a\1\s\src\Empactis.OrganisationManager.Data\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Data.dll
2017-10-18T14:23:02.6060314Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Commands\Empactis.OrganisationManager.Commands.csproj]
2017-10-18T14:23:05.6709386Z   Empactis.OrganisationManager.Commands -> d:\a\1\s\src\Empactis.OrganisationManager.Commands\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Commands.dll
2017-10-18T14:23:07.1448756Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj]
2017-10-18T14:23:10.6008560Z   Empactis.OrganisationManager -> d:\a\1\s\src\Empactis.OrganisationManager\bin\$(BuildConfiguration)\net462\win7-x86\Empactis.OrganisationManager.exe
2017-10-18T14:23:19.0356123Z   bower empactis-angular#1.14.0       not-cached https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:23:19.0364582Z   bower empactis-angular#1.14.0          resolve https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:23:19.9213747Z   bower empactis-angular#1.14.0         download https://github.com/empactis/empactis.angular/archive/1.14.0.tar.gz
2017-10-18T14:23:20.3151704Z   bower empactis-angular#1.14.0            retry Download of https://github.com/empactis/empactis.angular/archive/1.14.0.tar.gz failed with EHTTP, trying with git..
2017-10-18T14:23:20.3351529Z   bower empactis-angular#1.14.0         checkout 1.14.0
2017-10-18T14:23:21.9680144Z   bower empactis-angular#1.14.0         resolved https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:23:21.9680144Z   bower toaster#~0.4.14                  EINVRES Request to https://bower.herokuapp.com/packages/toaster failed with 410
2017-10-18T14:23:21.9850129Z d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj(91,5): error MSB3073: The command "bower install" exited with code 1.
2017-10-18T14:23:22.0610109Z ##[error]Error: C:\Program Files\dotnet\dotnet.exe failed with return code: 1
2017-10-18T14:23:22.0690403Z ##[error]Dotnet command failed with non-zero exit code on the following projects : d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj
2017-10-18T14:23:22.0860100Z ##[section]Finishing: dotnet publish
```

```
2017-10-18T14:32:12.3182581Z ##[section]Starting: dotnet publish
2017-10-18T14:32:12.3192572Z ==============================================================================
2017-10-18T14:32:12.3192572Z Task         : .NET Core
2017-10-18T14:32:12.3192572Z Description  : Build, test and publish using dotnet core command-line.
2017-10-18T14:32:12.3192572Z Version      : 1.0.2
2017-10-18T14:32:12.3192572Z Author       : Microsoft Corporation
2017-10-18T14:32:12.3192572Z Help         : [More Information](https://go.microsoft.com/fwlink/?linkid=832194)
2017-10-18T14:32:12.3192572Z ==============================================================================
2017-10-18T14:32:31.3345425Z [command]"C:\Program Files\dotnet\dotnet.exe" publish d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj --configuration $(BuildConfiguration) --output d:\a\1\a\Empactis.OrganisationManager
2017-10-18T14:32:31.7665616Z Microsoft (R) Build Engine version 15.1.1012.6693
2017-10-18T14:32:31.7665616Z Copyright (C) Microsoft Corporation. All rights reserved.
2017-10-18T14:32:31.7665616Z 
2017-10-18T14:32:34.7196822Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Shared\Empactis.OrganisationManager.Shared.csproj]
2017-10-18T14:32:36.8330692Z   Empactis.OrganisationManager.Shared -> d:\a\1\s\src\Empactis.OrganisationManager.Shared\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Shared.dll
2017-10-18T14:32:37.3147722Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Data\Empactis.OrganisationManager.Data.csproj]
2017-10-18T14:32:39.9288506Z   Empactis.OrganisationManager.Data -> d:\a\1\s\src\Empactis.OrganisationManager.Data\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Data.dll
2017-10-18T14:32:40.5508650Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager.Commands\Empactis.OrganisationManager.Commands.csproj]
2017-10-18T14:32:43.9879463Z   Empactis.OrganisationManager.Commands -> d:\a\1\s\src\Empactis.OrganisationManager.Commands\bin\$(BuildConfiguration)\net462\Empactis.OrganisationManager.Commands.dll
2017-10-18T14:32:45.9909878Z C:\Program Files\dotnet\sdk\1.0.4\Roslyn\Microsoft.CSharp.Core.targets(71,5): warning MSB3052: The parameter to the compiler is invalid, '/define:$(BUILDCONFIGURATION)' will be ignored. [d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj]
2017-10-18T14:32:49.5491470Z   Empactis.OrganisationManager -> d:\a\1\s\src\Empactis.OrganisationManager\bin\$(BuildConfiguration)\net462\win7-x86\Empactis.OrganisationManager.exe
2017-10-18T14:32:59.4380229Z   bower empactis-angular#1.14.0       not-cached https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:32:59.4390227Z   bower empactis-angular#1.14.0          resolve https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:33:00.3332745Z   bower empactis-angular#1.14.0         download https://github.com/empactis/empactis.angular/archive/1.14.0.tar.gz
2017-10-18T14:33:00.7847693Z   bower empactis-angular#1.14.0            retry Download of https://github.com/empactis/empactis.angular/archive/1.14.0.tar.gz failed with EHTTP, trying with git..
2017-10-18T14:33:00.7927707Z   bower empactis-angular#1.14.0         checkout 1.14.0
2017-10-18T14:33:03.7806194Z   bower empactis-angular#1.14.0         resolved https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0
2017-10-18T14:33:03.7846199Z   bower angular#1.5.9                    EINVRES Request to https://bower.herokuapp.com/packages/angular failed with 410
2017-10-18T14:33:03.8046195Z d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj(91,5): error MSB3073: The command "bower install" exited with code 1.
2017-10-18T14:33:03.8886238Z ##[error]Error: C:\Program Files\dotnet\dotnet.exe failed with return code: 1
2017-10-18T14:33:03.8896259Z ##[error]Dotnet command failed with non-zero exit code on the following projects : d:\a\1\s\src\Empactis.OrganisationManager\Empactis.OrganisationManager.csproj
2017-10-18T14:33:03.9096242Z ##[section]Finishing: dotnet publish
```

**Comment by bryanmacfarlane** [comment]
http 410 means the resource is gone.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/410 

Is the bower package (version) you are consuming gone?

Also, can you clarify where the url:

https://<redacted>:********@github.com/empactis/empactis.angular.git#1.14.0 

Is persisted?  in a file?  I redacted your PAT token from the post.

**Comment by tjrobinson** [comment]
@bryanmacfarlane Thanks, that was stupid of me - thanks for redacting it. I've re-generated the PAT as a precaution. I was in the middle of preparing for a live release so was in a hurry.

@alexmullans The URL `https://:********@github.com/empactis/empactis.angular.git#1.14.0` is stored in the `bower.json` file and is used to access the bower package `empactis.angular` from a private GitHub repository.

It looks like bower is trying to retrieve the package from `https://:********@github.com/empactis/empactis.angular.git#1.14.0` but this is failing (`Download of https://github.com/empactis/empactis.angular/archive/1.14.0.tar.gz failed with EHTTP, trying with git..`). Then the 410 error is because it looks like bower falls back to the main bower registry (`https://bower.herokuapp.com/`) which I would expect to fail as it's not a public package.

If this had never worked I would assume it's something I'd done but since it's been working with this setup for many months but only started failing today it feels like something is up on the VSTS side of things. Or GitHub itself.

This is the `bower.json`:

```
{
  "name": "orgman",
  "private": true,
  "dependencies": {
    "angular": "1.5.9",
    "angular-animate": "1.6.2",
    "angular-loading-bar": "~0.9.0",
    "angular-recursion": "~1.0.5",
    "angular-ui-select": "~0.19.4",
    "d3": "~3.5.6",
    "empactis-angular": "https://<redacted>:x-oauth-basic@github.com/empactis/empactis.angular.git#1.14.0",
    "jquery": "~2.2.4",
    "moment": "~2.13.0",
    "toaster": "~0.4.14",
    "ui-router": "~0.2.18"
  },
  "resolutions": {
    "angular": "1.5.9",
    "angular-animate": "1.6.2"
  }
}
```

**Comment by bryanmacfarlane** [comment]
https://github.com/empactis/empactis.angular/archive/1.14.0.tar.gz is a 404

I think this is between bower and github.

**Comment by alexmullans** [comment]
I agree with @bryanmacfarlane that this seems like an issue getting the content from GitHub. At the risk of asking a basic question, @tjrobinson with your regenerated PAT, does this same restore work locally if you [clear your Bower cache](https://bower.io/docs/api/#cache)?

**Comment by tjrobinson** [comment]
I've got it working now. I'm not sure exactly what fixed it but it must have been one of the following:

1. Changing the PAT to a new one (seems unlikely though as other builds using the old PAT had been fine)
2. Removing the `bower install` command from the `PrePublish` event (in the .csproj file) and using the 3rd party [Bower for Visual Studio Team Services](https://marketplace.visualstudio.com/items?itemName=touchify.vsts-bower) task instead - which uses the version of bower in the `node_modules` directory
3. Something was up with the network/GitHub/Bower but is now resolved through the passing of time

So I'm not exactly sure what fixed it, but I can close this issue now - thanks.
